servers/rendering/rendering_device_driver.h

```diff
@@ -0,0 +1,619 @@
+	virtual void shader_free(ShaderID p_shader) = 0;
+
+	/*********************/
+	/**** UNIFORM SET ****/
+	/*********************/
+
+	struct BoundUniform {
+		UniformType type = UNIFORM_TYPE_MAX;
+		uint32_t binding = 0xffffffff; // Binding index as specified in shader.
+		LocalVector<uint64_t> ids;
+	};
+
+	virtual UniformSetID uniform_set_create(VectorView<BoundUniform> p_uniforms, ShaderID p_shader, uint32_t p_set_index) = 0;
+	virtual void uniform_set_free(UniformSetID p_uniform_set) = 0;
+
+	/******************/
+	/**** TRANSFER ****/
+	/******************/
+
+	struct BufferCopyRegion {
+		uint64_t src_offset = 0;
+		uint64_t dst_offset = 0;
+		uint64_t size = 0;
+	};
+
+	struct TextureCopyRegion {
+		TextureSubresourceLayers src_subresources;
+		Vector3i src_offset;
+		TextureSubresourceLayers dst_subresources;
+		Vector3i dst_offset;
+		Vector3i size;
+	};
+
+	struct BufferTextureCopyRegion {
+		uint64_t buffer_offset = 0;
+		uint32_t buffer_row_length = 0;
+		uint32_t buffer_texture_height = 0;
+		TextureSubresourceLayers texture_subresources;
+		Vector3i texture_offset;
+		Vector3i texture_region_size;
+	};
+
+	enum TextureCopyMode {
+		TEXTURE_COPY_MODE_COPY,
+		TEXTURE_COPY_MODE_RESOLVE,
+	};
+
+	virtual void command_clear_buffer(CommandBufferID p_cmd_buffer, BufferID p_buffer, uint64_t p_offset, uint64_t p_size) = 0;
+	virtual void command_copy_buffer(CommandBufferID p_cmd_buffer, BufferID p_src_buffer, BufferID p_dst_buffer, VectorView<BufferCopyRegion> p_regions) = 0;
+	virtual void command_copy_texture(CommandBufferID p_cmd_buffer, TextureID p_src_texture, TextureLayout p_src_texture_layout, TextureID p_dst_texture, TextureLayout p_dst_texture_layout, VectorView<TextureCopyRegion> p_regions, TextureCopyMode p_mode) = 0;
+
+	virtual void command_clear_color_texture(CommandBufferID p_cmd_buffer, TextureID p_texture, TextureLayout p_texture_layout, const Color &p_color, const TextureSubresourceRange &p_subresources) = 0;
+
+	virtual void command_copy_buffer_to_texture(CommandBufferID p_cmd_buffer, BufferID p_src_buffer, TextureID p_dst_texture, TextureLayout p_dst_texture_layout, VectorView<BufferTextureCopyRegion> p_regions) = 0;
+	virtual void command_copy_texture_to_buffer(CommandBufferID p_cmd_buffer, TextureID p_src_texture, TextureLayout p_src_texture_layout, BufferID p_dst_buffer, VectorView<BufferTextureCopyRegion> p_regions) = 0;
+
+	/******************/
+	/**** PIPELINE ****/
+	/******************/
+
+	virtual void pipeline_free(PipelineID p_pipeline_id) = 0;
+
+	// ----- BINDING -----
+
+	virtual void command_bind_push_constants(CommandBufferID p_cmd_buffer, ShaderID p_shader, uint32_t p_offset, VectorView<uint32_t> p_data) = 0;
+
+	// ----- CACHE -----
+
+	virtual bool pipeline_cache_create(const Vector<uint8_t> &p_data) = 0;
+	virtual void pipeline_cache_free() = 0;
+	virtual size_t pipeline_cache_query_size() = 0;
+	virtual Vector<Vector<uint8_t>> pipeline_cache_serialize() = 0;
```

Comment:
I'm not very concerned about the allocations, since, after all, that's being called when about to hit the disk. Also, one of the chunks has actually been already allocated by the driver.

Regarding the obscurity of the return type, I can't think of a good one. It's just a number of vectors the caller is expected to write back-to-back in the cache file.

Reply:
Why is it a Vector of Vectors but the input is a Vector though? Am I supposed to concatenate them and then use them back again as the input?

Reply:
That _was_ the idea. And I'm saying _was_ because on closer inspection I've realized I can get rid of that ugliness. Next push will come with a pain `Vector<uint8_t>` as the return type.